### PR TITLE
RavenDB-16303: Add conversion of GUID to UUID in RavenDB Postgresql ETL

### PR DESCRIPTION
**Pull request by garayx**

### CODE REVIEW

File: src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs [hunk 1]
```diff
@@ -469,7 +470,16 @@ public static void SetParamValue(DbParameter colParam, SqlColumn column, List<Fu
                                 objectValue.TryGetMember(nameof(SqlDocumentTransformer.VarcharFunctionCall.Value), out object fieldValue))
                             {
                                 colParam.DbType = (DbType)Enum.Parse(typeof(DbType), dbType.ToString(), false);
-                                colParam.Value = fieldValue.ToString();
+
+                                var value = fieldValue.ToString();
+                                if (colParam.DbType == DbType.Guid && Guid.TryParse(value, out var guid) && colParam is Npgsql.NpgsqlParameter p)
```
Comment (ayende) [inline src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs]: Why do we care this is `NpgsqlParameter ` ? Shouldn't this be broadly applicable?

File: src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs [hunk 2]
```diff
@@ -561,6 +571,28 @@ private static void SetParamStringValue(DbParameter colParam, string value, List
...
+                            {
+                                p.Value = guid;
+                                p.NpgsqlDbType = NpgsqlDbType.Uuid;
+                                return true;
+                            }
+
+                            break;
+                        default:
+                            return false;
+                    }
```
Comment (ayende) [inline src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs]: Why just for PSGL?

File: test/SlowTests/Issues/RavenDB_16303.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+{
+    public class RavenDB_16303 : EtlTestBase
+    {
+        public RavenDB_16303(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [InlineData("{ Value: this.Guid, Type: 'Guid' }")]
+        [InlineData("this.Guid")]
+        [RequiresNpgSqlInlineData]
```
Comment (arekpalinski) [inline test/SlowTests/Issues/RavenDB_16303.cs:27]: If this change won't be Postgresql specific (see Oren's comments) then we can drop `RequiresNpgSqlInlineData`, right?

Comment (garayx) [reply test/SlowTests/Issues/RavenDB_16303.cs:27]: no the connection string (line 30) and maybe the sql would have to be changed

File: src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs (hunk 1 above)
Comment (garayx) [reply src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs]: I didn't test it on other dbs, so cannot be sure, do you want to fix it here or we can open an new issue for it ?

Comment (arekpalinski) [reply src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs]: We need to know if this issue is specific only for PSGL or we need that for other relational dbs as well

Comment (arekpalinski) [reply src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs]: At glance it looks this is not only PSGL related

**Comment by arekpalinski** [comment]
https://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.2/4357/consoleFull

I see that some of added test cases have been skipped:

```
17:43:24      SlowTests.Issues.RavenDB_16303.CanCreateSqlEtlTransformScriptWithGuidWhenTableTypeGuid(provider: MySQL)
17:43:24      SlowTests.Issues.RavenDB_16303.CanCreateSqlEtlTransformScriptWithGuidWhenTableTypeGuid(provider: MySQL) [SKIP]
17:43:24        Test requires MySQL database
17:43:24      SlowTests.Issues.RavenDB_16303.CanCreateSqlEtlTransformScriptWithGuidWhenTableTypeGuid(provider: NpgSQL)
17:43:24      SlowTests.Issues.RavenDB_16303.CanCreateSqlEtlTransformScriptWithGuidWhenTableTypeGuid(provider: NpgSQL) [SKIP]
17:43:24        Test requires NpgSQL database
17:43:24      SlowTests.Issues.RavenDB_16303.CanCreateSqlEtlTransformScriptWithGuidWhenTableTypeGuid(provider: Oracle)
17:43:24      SlowTests.Issues.RavenDB_16303.CanCreateSqlEtlTransformScriptWithGuidWhenTableTypeGuid(provider: Oracle) [SKIP]
17:43:24        Test requires Oracle database
```

only the one for MSSQL has passed:

`SlowTests.Issues.RavenDB_16303.CanCreateSqlEtlTransformScriptWithGuidWhenTableTypeGuid(provider: MsSQL)`

Don't we have it properly configured?

CC @topalach

**Comment by topalach** [comment]
We have the following environment variables set up in jenkins tests: `RAVEN_MONGODB_CONNECTION_STRING`, `RAVEN_MSSQL_CONNECTION_STRING`, `RAVEN_MYSQL_CONNECTION_STRING`, `RAVEN_NPGSQL_CONNECTION_STRING`, `RAVEN_ORACLESQL_CONNECTION_STRING`.

**Review by ppekrol — APPROVED** [review]

**Comment by arekpalinski** [comment]
I have scheduled another test build: https://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.2/4362/

**Comment by arekpalinski** [comment]
4 tests have failed there

File: src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs [hunk 4]
```diff
@@ -468,8 +471,43 @@ public static void SetParamValue(DbParameter colParam, SqlColumn column, List<Fu
                             if (objectValue.TryGetMember(nameof(SqlDocumentTransformer.VarcharFunctionCall.Type), out object dbType) &&
                                 objectValue.TryGetMember(nameof(SqlDocumentTransformer.VarcharFunctionCall.Value), out object fieldValue))
                             {
-                                colParam.DbType = (DbType)Enum.Parse(typeof(DbType), dbType.ToString(), false);
-                                colParam.Value = fieldValue.ToString();
+                                try
+                                {
+                                    colParam.DbType = (DbType)Enum.Parse(typeof(DbType), dbType.ToString(), false);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs]: Use TryParse?
BTW. Do we want to be case sensitive here?

File: src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs [hunk 5]
```diff
@@ -468,8 +471,43 @@ public static void SetParamValue(DbParameter colParam, SqlColumn column, List<Fu
...
+                                } catch
+                                {
+                                    // keep default
+                                    colParam.DbType = DbType.String;
+                                }
+
+                                var value = fieldValue.ToString();
+                                if ((colParam.DbType == DbType.Guid || colParam.DbType == DbType.String) && Guid.TryParse(value, out var guid))
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs]: Why DbType.String?

File: src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/DbProviderFactories.cs [hunk 6]
```diff
@@ -38,7 +38,7 @@ public static DbCommandBuilder InitializeCommandBuilder(this DbProviderFactory f
                     Start = "[",
                     End = "]"
                 };
-            if (factory is NpgsqlFactory)
+            if (factory is NpgsqlFactory || factory is OracleClientFactory)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/DbProviderFactories.cs:41]: Is it a breaking change? Why we never run into this before?

File: src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs (hunk 4 above)
Comment (garayx) [reply src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/RelationalDatabaseWriter.cs]: The `DbType' setter of OracleParameter throws, not the enum
image

Yes, this is a case when you pass the type explicitly

File: src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/DbProviderFactories.cs (hunk 6 above)
Comment (garayx) [reply src/Raven.Server/Documents/ETL/Providers/SQL/RelationalWriters/DbProviderFactories.cs:41]: no, its used for proper SQL script created by our ETL, I think there are no tests for oracle etl.

**Pull request merged**